lib/open_food_network/order_cycle_permissions.rb

```diff
@@ -252,18 +206,54 @@ def editable_variants_for_outgoing_exchanges_to(hub)
         granting_producer_ids = related_enterprises_granting(:add_to_order_cycle,
                                                              to: [hub.id],
                                                              scope: granted_producers)
-        permitted_variants = Spree::Variant.joins(:product).
-          where('spree_products.supplier_id IN (?)', granting_producer_ids)
+        permitted_variants = variants_from_suppliers(granting_producer_ids)
 
         Spree::Variant.where(id: permitted_variants)
       end
     end
 
     private
 
+    def visible_and_editable_variants(hub)
+      # Producers that have granted the hub P-OC
+      producer_ids = related_enterprises_granting(:add_to_order_cycle,
+                                                  to: [hub.id],
+                                                  scope: Enterprise.is_primary_producer)
+
+      # Variants from Producers via permissions, and from the hub itself
+      available_variants = variants_from_suppliers(producer_ids.push(hub.id))
+
+      # PLUS variants that are already in an outgoing exchange of this hub, so things don't break
+      active_variants = active_outgoing_variants(hub)
+
+      Spree::Variant.where(id: available_variants | active_variants)
+    end
+
+    def variants_from_suppliers(supplier_ids)
+      Spree::Variant.joins(:product).where('spree_products.supplier_id IN (?)', supplier_ids)
+    end
+
+    def active_outgoing_variants(hub)
+      @active_outgoing_variants ||= begin
+        active_variants = []
+        @order_cycle.exchanges.outgoing.where(receiver_id: hub).limit(1).each do |exchange|
+          active_variants = exchange.variants
+        end
+        active_variants
+      end
```

Comment:
there'a a limit(1) in there. it means first_exchange, right?

```suggestion
     def active_outgoing_variants(hub)
      @active_outgoing_variants ||= begin
        
        first_outgoing_exchange = @order_cycle.exchanges.outgoing.where(receiver_id: hub).limit(1)
        if first_outgoing_exchange.present?
           first_outgoing_exchange.variants            
        else
              []
        end
      end
```

Reply:
I think we can simplify this further, into a one-liner:
```ruby
@order_cycle.exchanges.outgoing.where(receiver_id: hub).first.andand.variants || []
```